# Lab: Building a Reproducible Analysis Workflow

In the first part of this lab, you will use official National Park Service data to answer a guided question. In the second part, you will find a dataset and conduct an analysis of your own.

Use the workflow from lecture throughout the lab:

1. **Question**
2. **Data**
3. **Operation**
4. **Check**
5. **Evidence**
6. **Conclusion**
7. **Limitation**


## Learning goals

By the end of this lab, you should be able to:

- organize an investigation with the reproducible analysis workflow;
- inspect a dataset before analyzing it;
- group, summarize, check, and sort observations with Pandas;
- formulate a research question that can be answered with data; and
- distinguish evidence, conclusions, and limitations.


## Using AI tools responsibly

You may use AI to ask questions, debug code, or receive feedback. You remain responsible for understanding your work, verifying suggestions against the data, and citing the source of your dataset. At the end, disclose whether you used an AI tool. Using AI is not required.


# Part 1: Florida park visitation


## Background

The National Park Service (NPS) publishes estimates of recreational visits to its units. These estimates help parks understand patterns in public use and plan staffing, maintenance, transportation, and visitor services.

This course file contains monthly estimates for five Florida NPS units from 2015 through 2025:

- Big Cypress National Preserve
- Biscayne National Park
- Canaveral National Seashore
- Dry Tortugas National Park
- Everglades National Park

A recreational visit is a visit, not necessarily one unique person. A person who visits several times may be counted several times.


| Feature | Description |
| --- | --- |
| `park_code` | Four-letter NPS unit code |
| `park` | Name of the NPS unit |
| `year` | Calendar year of the observation |
| `month_number` | Month represented as an integer from 1 through 12 |
| `month` | Month name |
| `recreation_visits` | Estimated recreational visits during that park-month |


## Step 1: Question

**After the decline in July 2020, when did Canaveral National Seashore's July recreational visitation first return to or exceed its average July visitation from 2015–2019?**

The 2015–2019 mean provides a pre-pandemic comparison point. In this question, “recovery” means the first July from 2020 onward with visitation at or above that baseline. This definition does not require every later July to remain above the baseline.


## Step 2: Data

Import Pandas using the alias `pd`. Read `data/nps_florida_park_visitation_2015_2025.csv` into a DataFrame named `park_visits`, then display its first five rows.


In [59]:
# Write your code here.
import pandas as pd

park_visits = pd.read_csv("data/nps_florida_park_visitation_2015_2025.csv")

park_visits.head()

,park_code,park,year,month_number,month,recreation_visits
0,BICY,Big Cypress National Preserve,2015,1,January,114045
1,BICY,Big Cypress National Preserve,2015,2,February,144112
2,BICY,Big Cypress National Preserve,2015,3,March,153084
3,BICY,Big Cypress National Preserve,2015,4,April,113880
4,BICY,Big Cypress National Preserve,2015,5,May,72510


Display the last five rows.


In [60]:
# Write your code here.
park_visits.tail()

,park_code,park,year,month_number,month,recreation_visits
655,EVER,Everglades National Park,2025,8,August,50464
656,EVER,Everglades National Park,2025,9,September,32290
657,EVER,Everglades National Park,2025,10,October,19008
658,EVER,Everglades National Park,2025,11,November,53416
659,EVER,Everglades National Park,2025,12,December,70926


Display information about the DataFrame and calculate the number of missing observations in each feature.


In [61]:
# Write your code here.
park_visits.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660 entries, 0 to 659
Data columns (total 6 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   park_code          660 non-null    object
 1   park               660 non-null    object
 2   year               660 non-null    int64 
 3   month_number       660 non-null    int64 
 4   month              660 non-null    object
 5   recreation_visits  660 non-null    int64 
dtypes: int64(3), object(3)
memory usage: 31.1+ KB


In [62]:
park_visits.isnull().sum()

park_code            0
park                 0
year                 0
month_number         0
month                0
recreation_visits    0
dtype: int64

**What does each row represent?**

Your answer: The park_code row is a string/object showing the park identifier, the park is also a string/object showing the park name, year is an int64, month_number is an int64, month is a string/object with the actual name of the month corresponding to the month_number, and hte recreation_visits is an int64 showing the number of visits to that park that month of that year


**Are any observations missing from the six features?**

Your answer: No observations are missing


## Step 3: Operation

Create one Boolean condition that identifies Canaveral National Seashore and another that identifies July.


In [63]:
# Write your code here.

cns = (park_visits[ 'park_code' ] == "CANA" )

jul = (park_visits[ 'month' ] == "July" )


Combine the two conditions with `&`. Use `.loc` to select the `year` and `recreation_visits` features, make a copy named `canaveral_july`, and display it.


In [64]:
# Write your code here.
canaveral_july = park_visits.loc()[cns & jul, ['year', 'recreation_visits']].copy()

canaveral_july

,year,recreation_visits
270,2015,163169
282,2016,216437
294,2017,138018
306,2018,189285
318,2019,162947
330,2020,112898
342,2021,206636
354,2022,233352
366,2023,173296
378,2024,176242


Create `pre_pandemic_condition`, which is `True` for years from 2015 through 2019. Use it with `.loc` to calculate the mean of `recreation_visits`, and store the result as `pre_pandemic_july_mean`.

Display `pre_pandemic_july_mean`.


In [65]:
# Write your code here.
pre_pandemic_condition = canaveral_july['year'] <= 2019

pre_pandemic_july_mean = canaveral_july.loc()[pre_pandemic_condition, 'recreation_visits'].mean()

pre_pandemic_july_mean

# ANSWER = 173971.2

np.float64(173971.2)

Create a condition that identifies rows from 2020 onward whose `recreation_visits` are greater than or equal to `pre_pandemic_july_mean`. Use `.loc` to create `recovery_years`, then display it in ascending year order.


In [66]:
# Write your code here.
recovery_condition = ((canaveral_july['year'] >= 2020) & (canaveral_july['recreation_visits'] >= pre_pandemic_july_mean))

recovery_years = canaveral_july.loc()[recovery_condition, ['year', 'recreation_visits']].sort_values(by='year', ascending=True)

recovery_years

,year,recreation_visits
342,2021,206636
354,2022,233352
378,2024,176242
390,2025,253403


## Step 4: Check

Display the number of rows in `canaveral_july`, its smallest year, and its greatest year.


In [67]:
# Write your code here.
canaveral_july['year'].count()

np.int64(11)

In [68]:
canaveral_july['year'].min()

2015

In [69]:
canaveral_july['year'].max()

2025

**Do these results confirm that the filtered data contains one July observation for every year from 2015 through 2025?**

Your answer: Yes, there are 11 observations spanning from 2015 - 2025, one for each year.


Count the observations used to calculate `pre_pandemic_july_mean`.


In [70]:
# Write your code here.
canaveral_july.loc()[pre_pandemic_condition, 'recreation_visits'].count()

np.int64(5)

**Does the baseline use the expected five years?**

Your answer: Yes, it uses the expected five years.


## Step 5: Evidence

Display the first row of `recovery_years`.


In [71]:
# Write your code here.
recovery_years.head(1)

,year,recreation_visits
342,2021,206636


**What was the 2015–2019 mean, and which year first met or exceeded it after the July 2020 decline? Report that year's July visitation.**

Your answer: The pre_pandemic_july_mean was 173971.2, and the first year to meet or exceend it was in 2021, with 206636 visits in July.


## Step 6: Conclusion


**Write one sentence that answers the research question using the evidence.**

Your answer: The first year after the pandemic-caused decline in 2020 to meet or exceed 
             the average July visitation for 2015-2019 was July 2021, with 206,636 visits.

## Step 7: Limitation


**Does first exceeding the baseline mean that every later July remained above it? Use the data to explain.**

Your answer: No it does not. The data shows that July 2023 did not meet or exceed the 2015-2019 average, suggesting a surge after parks reopened that settled down. July visits are not required to stay above the average, they continue to contribute to the average.


**Why does this analysis alone not establish that the pandemic caused the 2020 decline?**

Your answer: This analyis alone does not establish that the decline was caused by the pandemic simply because it is July 2023 that is below the average from 2015-2019. Since 2021 and 2022 were above the average, 2023's lower number doesn't necessarily relate to the pandemic that was now 3 years removed.


# Part 2: Your own investigation


## Find a dataset

Find a CSV dataset that interests you. Kaggle is one possible source, but you may use another reputable source. Choose a dataset that:

- has a clearly identified publisher or creator;
- includes documentation explaining the observations and features;
- contains at least one categorical and one quantitative feature;
- is appropriate to share in a class assignment;
- does not contain private or personally identifying information; and
- is small enough to run comfortably in this notebook.

Save the CSV in the notebook's `data` folder before continuing.


**What is the dataset's title, publisher or creator, and direct source URL?**

Your answer: "Online Food Ordering Dataset", Srimathy Sivanessan, https://www.kaggle.com/datasets/srisyra02/online-food-ordering-dataset?resource=download


**What does one row represent?**

Your answer: Shows the Age, Gender, Marital Status, Occupation, Monthly Income, Educational Qualifications, Family Size, Customer Type, Latitude, Longitude, Pin Code, Output, Feedback 


## Step 1: Question

Formulate one research question that can be answered with your dataset. The question must require a calculation or comparison, not merely looking up one value.


**What is your research question?**

Your answer: Do customers who give positive feedback have a different average family size than customers who give negative feedback?


## Step 2: Data

Read your CSV into a clearly named DataFrame. Display the first and last five rows, information about the DataFrame, and the missing-value count for every feature.


In [72]:
# Write your code here.
food_delivery = pd.read_csv("data/online food delivery dataset.csv")

food_delivery.head()

,Age,Gender,Marital Status,Occupation,Monthly Income,Educational Qualifications,Family size,Customer Type,latitude,longitude,Pin code,Output,Feedback,Unnamed: 13
0,20,Female,Single,Student,No Income,Post Graduate,4,Frequent,12.9766,77.5993,560001,Yes,Positive,Yes
1,24,Female,Single,Student,Below Rs.10000,Graduate,3,Regular,12.9770,77.5773,560009,Yes,Positive,Yes
2,22,Male,Single,Student,Below Rs.10000,Post Graduate,3,Regular,12.9551,77.6593,560017,Yes,Negative,Yes
3,22,Female,Single,Student,No Income,Graduate,6,Frequent,12.9473,77.5616,560019,Yes,Positive,Yes
4,22,Male,Single,Student,Below Rs.10000,Post Graduate,4,Frequent,12.9850,77.5533,560010,Yes,Positive,Yes


In [73]:
food_delivery.tail()

,Age,Gender,Marital Status,Occupation,Monthly Income,Educational Qualifications,Family size,Customer Type,latitude,longitude,Pin code,Output,Feedback,Unnamed: 13
383,23,Female,Single,Student,No Income,Post Graduate,2,Regular,12.9766,77.5993,560001,Yes,Positive,Yes
384,23,Female,Single,Student,No Income,Post Graduate,4,Frequent,12.9854,77.7081,560048,Yes,Positive,Yes
385,22,Female,Single,Student,No Income,Post Graduate,5,Frequent,12.9850,77.5533,560010,Yes,Positive,Yes
386,23,Male,Single,Student,Below Rs.10000,Post Graduate,2,Regular,12.9770,77.5773,560009,Yes,Positive,Yes
387,23,Male,Single,Student,No Income,Post Graduate,5,Frequent,12.8988,77.5764,560078,Yes,Positive,Yes


In [74]:
food_delivery.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 388 entries, 0 to 387
Data columns (total 14 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Age                         388 non-null    int64  
 1   Gender                      388 non-null    object 
 2   Marital Status              388 non-null    object 
 3   Occupation                  388 non-null    object 
 4   Monthly Income              388 non-null    object 
 5   Educational Qualifications  388 non-null    object 
 6   Family size                 388 non-null    int64  
 7   Customer Type               388 non-null    object 
 8   latitude                    388 non-null    float64
 9   longitude                   388 non-null    float64
 10  Pin code                    388 non-null    int64  
 11  Output                      388 non-null    object 
 12  Feedback                    388 non-null    object 
 13  Unnamed: 13                 388 non

In [75]:
food_delivery.isnull().sum()

Age                           0
Gender                        0
Marital Status                0
Occupation                    0
Monthly Income                0
Educational Qualifications    0
Family size                   0
Customer Type                 0
latitude                      0
longitude                     0
Pin code                      0
Output                        0
Feedback                      0
Unnamed: 13                   0
dtype: int64

**Which observations and features will you use to answer your question?**

Your answer: Family size, Feedback


## Step 3: Operation

Write the code needed to prepare, filter, group, summarize, or sort the data. Store the final evidence in a clearly named object and display it.


In [76]:
# Write your code here.
feedback_familySize = food_delivery.groupby('Feedback')['Family size'].mean()
feedback_familySize

Feedback
Negative     3.309859
Positive     3.274448
Name: Family size, dtype: float64

**Explain how your operations connect the selected data to your research question.**

Your answer: I grouped the observations by their Feedback responses and calculated the mean Family size for each group. This let me compare the average family size of customers who left positive or negative feedback. The results show very little difference between them, with customers giving negative feedback having a slightly larger family size (3.31) than those that gave positive feedback (3.27).

## Step 4: Check

Write at least one check that could reveal an error in your analysis, then display its result.


In [77]:
# Write your code here.
food_delivery['Feedback'].value_counts()


Feedback
Positive     317
Negative      71
Name: count, dtype: int64

**What did you check, and did the result meet your expectation?**

Your answer: I checked the number of observations in each Feedback column using value_counts(). I expected the Positive and Negative categories to contain observations because my analysis compared the average family size between the two. With the results being 317 Positive and 71 Negative, both groups contain data and meet my expectations.


## Step 5: Evidence


**Which specific value or comparison in your output answers the question?**

Your answer: The average family size for negative feedback providers (3.31) is slightly larger than the average family size for positive feedback providers (3.27). My question asked whether those giving positive vs negative feedback had different average family sizes, and this comparison answers that question.

## Step 6: Conclusion


**What conclusion is supported by the evidence?**

Your answer: There is no major difference in positive vs negative feedback providers based on average family size.


## Step 7: Limitation


**What should a reader avoid concluding from your analysis, and why?**

Your answer: The dataset only contains 388 observations and the ages are all on the younger spectrum. Readers should avoid drawing any firm conclusions from this data as representative of the whole population.


## AI-use reflection


**State whether you used an AI tool. If you did, describe one suggestion you checked and what you accepted, changed, or rejected. If you did not, state that you did not use one.**

Your response: I used ChatGPT to figure out why my original code wasn't working for Step 3 when I was creating the canaveral_july copy (I had & and .loc in different places, causing the error).


## Submission checklist

Before submitting, confirm that all cells run in order, the independent dataset and its source are included, every answer is complete, numerical claims match the output, conclusions are appropriately limited, and the AI-use reflection is complete.
